File: src/sliding-window/moving-average-from-data-stream.ts
// DIFFICULTY: EASY
//
// Given a stream of integers and a window size, calculate the moving average of all integers in the sliding window.
//
// Implement the MovingAverage class:
//
// - MovingAverage(int size) Initializes the object with the size of the window size.
// - double next(int val) Returns the moving average of the last size values of the stream.
//
// See {@link https://leetcode.com/problems/moving-average-from-data-stream/}
export { MovingAverage };

// SOLUTION:
//
// This can be solved with a queue or sliding window.
class MovingAverage {
  private readonly size: number;
  private readonly window: number[];
  private sum: number;

  constructor(size: number) {
    this.size = size;
    this.window = [];
    this.sum = 0;
  }

  next(val: number): number {
    if (this.window.length === this.size) {
      const out = this.window.shift();
      this.sum -= out!;
    }

    this.window.push(val);
    this.sum += val;
    return this.sum / this.window.length;
  }
}

File: src/leetcode/sliding_window/moving_average_from_data_stream.py
# DIFFICULTY: EASY
# ----------------
#
# Given a stream of integers and a window size, calculate the moving average of all integers in the sliding window.
#
# Implement the MovingAverage class:
#
# - MovingAverage(int size) Initializes the object with the size of the window size.
# - double next(int val) Returns the moving average of the last size values of the stream.
#
# See https://leetcode.com/problems/moving-average-from-data-stream
from collections import deque


class MovingAverage:
    """
    SOLUTION
    --------

    This can be solved with a queue or sliding window.

    COMPLEXITY
    ----------

    Time complexity is O(1) for next.

    Space complexity is O(n) where n is the size of the window.
    """

    def __init__(self, size: int) -> None:
        self.size = size
        self.window: deque[int] = deque()
        self.total = 0

    def next(self, val: int) -> float:
        if len(self.window) == self.size:
            self.total -= self.window.popleft()

        self.window.append(val)
        self.total += val
        return self.total / len(self.window)
